### DISCUSSION TITLE
Is it possible to insert auto ads between paragraphs

### DISCUSSION THREAD

**Discussion by robertmain53** [11ty/eleventy#2355] [Q&A]
Let's say that I want to build a magazine website with 11ty and monetize it with ads. 
In particular I would like to add ads after paragraph "x" and paragraph "y" of each piece of content.

Content is created in markdown files. 

Is it possible to achieve this result in 11ty without manually insert a shortcode?

**Comment by pdehaan** [comment]
I can't immediate solution apart from a shortcode or some sort of manual placeholder HTML comment like `<!-- ad -->` that you replace at build time using a transform.

Unless you could possibly reparse the DOM using something like https://cheerio.js.org/ in a transform to determine whatever your paragraph "x" and "y" are and inject the code.
It could be tricky since if it's always inserting ads after the, say, 3rd and 6th `<p>` tag in a post, you could have some edge cases like ads being injected after a `<p>` but before a `<pre>` code block which might look strange.

**Reply by robertmain53** [reply]
Thank you, i ill look into cheerio. The placeholder solution is not scalable when one have i.e. 50 paragraphs and content that gets updaded often.

**Reply by pdehaan** [reply]
@robertmain53 How were you thinking of injecting ads?
> I would like to add ads after paragraph "x" and paragraph "y" of each piece of content.

Were `x` and `y` like, the third and sixth `<p>` tag in an article? Or did you have some other idea of where you ideally wanted to modify the DOM and add ads (like, the second paragraph after some `<h2>` heading tag)?

**Reply by robertmain53** [reply]
The first one. So, for instance, after paragraphs: 2, 8, 16, 24, 32, 40. In this case there could be up to 6 different ads to add.

**Reply by pdehaan** [reply]
Cool. OK, this seemed to work for my simple test case:

```js
// .eleventy.js
const cheerio = require("cheerio");

module.exports = function (eleventyConfig) {
  eleventyConfig.addTransform("inject-ads", function (content) {
    if (!this.outputPath?.endsWith(".html")) {
      console.log("abort early:", this.outputPath);
      return content;
    }
    const $ = cheerio.load(content);
    // Inject ads after 2nd <p> tag, then after 8,16,24,32, etc.
    const r = $("main p:nth-of-type(2), main p:nth-of-type(8n)");
    $(r).after("<!-- AD HERE -->");
    return $.html();
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

Still a lot if imaginary sharp edges probably.
For example, this worked for my simple test case w/ 1 file, but there isn't currently a way to limit this to certain files/paths. So if I had an imaginary homepage or about page, it'd probably inject a bunch of `<!-- AD -->` markers in those files too.
I imagine you could add some other `data-no-ads` attribute on pages you may not want ads injected into.
Or maybe be more specific with your selectors, like `main.page-w-ads p:nth-of-type(2)`.
Or possibly check the `this.outputPath` for some other path/slug like `/blog/` before injecting ads with reckless abandon.

But here was my test file (creates a quick ~45 paragraphs):
```md
---
title: HoMe PaGe
layout: base.liquid
---

## {{ title }}

{% for idx in (1..45) %}
paragraph {{ idx }}
{% endfor %}
```

And my output looks generally like:

```html
<main>
  <h2>HoMe PaGe</h2>
  <p>paragraph 1</p>
  <p>paragraph 2</p>
  <!-- AD HERE -->
  <p>paragraph 3</p>
  <p>paragraph 4</p>
  <p>paragraph 5</p>
  <p>paragraph 6</p>
  <p>paragraph 7</p>
  <p>paragraph 8</p>
  <!-- AD HERE -->
  <p>paragraph 9</p>
  <p>paragraph 10</p>
  <p>paragraph 11</p>
  <p>paragraph 12</p>
  <p>paragraph 13</p>
  <p>paragraph 14</p>
  <p>paragraph 15</p>
  <p>paragraph 16</p>
  <!-- AD HERE -->
  <p>paragraph 17</p>
  <p>paragraph 18</p>
  ...
</main>
```

**Reply by pdehaan** [reply]
Sample repo at https://github.com/pdehaan/11ty-2355

**Reply by robertmain53** [reply]
Very good. This works. All details (how to hide, where to show, etc.) can be easily fixed, no problem,
Thank you a lot
